src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -125,6 +125,10 @@ public FileStorage(final Path path) {
 
     @Override
     public CompletableFuture<Void> save(final Key key, final Content content) {
+        final Path next = this.dir.resolve(key.string());
+        if (!next.normalize().startsWith(next)) {
+            throw new ArtipieException("Entry path is out of storage.");
```

Comment:
This exception should be put as the result of `CompletableFuture`, you can use helper object for Java-1.8 [FailedCompletionStage](https://github.com/artipie/asto/blob/master/src/main/java/com/artipie/asto/FailedCompletionStage.java):
```java
return new FailedCompletionStage<>(new ArtipieException("Entry path is out of storage.")).toCompletableFuture();
```

Also, I'd suggest using [ArtipieIOException](https://github.com/artipie/asto/blob/master/src/main/java/com/artipie/asto/ArtipieIOException.java) here (since it's IO operation error).

Reply:
@g4s8 got it